File: android/app/src/main/java/com/gastosapp/AddExpenseActivity.java
package com.gastosapp;

import android.Manifest;
import android.app.DatePickerDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;

import com.gastosapp.data.local.ExpenseEntity;
import com.gastosapp.viewmodel.ExpenseViewModel;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Locale;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class AddExpenseActivity extends AppCompatActivity {

    private EditText etDescription;
    private EditText etAmount;
    private Spinner spinnerCategory;
    private EditText etDate;
    private Button btnSave;
    private ImageButton btnMic;
    private SpeechRecognizer speechRecognizer;
    private boolean isLoading = false;
    private boolean isListening = false;
    private Calendar selectedDate;
    private ExpenseViewModel viewModel;

    private static final int REQUEST_RECORD_AUDIO_PERMISSION = 200;

    private static final String[] CATEGORIES = {
            "Selecciona una categoría",
            "Alimentos",
            "Transporte",
            "Entretenimiento",
            "Salud",
            "Servicios",
            "Otros"
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_expense);

        viewModel = new ViewModelProvider(this).get(ExpenseViewModel.class);

        initViews();
        setupCategorySpinner();
        setupDatePicker();
    }

    private void initViews() {
        ImageButton btnBack = findViewById(R.id.btnBack);
        etDescription = findViewById(R.id.etDescription);
        etAmount = findViewById(R.id.etAmount);
        spinnerCategory = findViewById(R.id.spinnerCategory);
        etDate = findViewById(R.id.etDate);
        btnSave = findViewById(R.id.btnSave);
        btnMic = findViewById(R.id.btnMic);

        btnBack.setOnClickListener(v -> finish());
        btnSave.setOnClickListener(v -> handleSave());
        btnMic.setOnClickListener(v -> toggleVoiceRecognition());

        initSpeechRecognizer();

        // Set default date to today
        selectedDate = Calendar.getInstance();
        updateDateDisplay();
    }

    private void setupCategorySpinner() {
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                CATEGORIES);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinnerCategory.setAdapter(adapter);
    }

    private void setupDatePicker() {
        etDate.setOnClickListener(v -> {
            DatePickerDialog datePickerDialog = new DatePickerDialog(
                    this,
                    (view, year, month, dayOfMonth) -> {
                        selectedDate.set(year, month, dayOfMonth);
                        updateDateDisplay();
                    },
                    selectedDate.get(Calendar.YEAR),
                    selectedDate.get(Calendar.MONTH),
                    selectedDate.get(Calendar.DAY_OF_MONTH));
            datePickerDialog.show();
        });
        etDate.setFocusable(false);
    }

    private void updateDateDisplay() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        etDate.setText(sdf.format(selectedDate.getTime()));
    }

    private void handleSave() {
        String description = etDescription.getText().toString().trim();
        String amountStr = etAmount.getText().toString().trim();
        int categoryPosition = spinnerCategory.getSelectedItemPosition();

        if (description.isEmpty()) {
            Toast.makeText(this, "Por favor ingresa una descripción", Toast.LENGTH_SHORT).show();
            return;
        }

        if (amountStr.isEmpty()) {
            Toast.makeText(this, "Por favor ingresa un monto", Toast.LENGTH_SHORT).show();
            return;
        }

        if (categoryPosition == 0) {
            Toast.makeText(this, "Por favor selecciona una categoría", Toast.LENGTH_SHORT).show();
            return;
        }

        double amount;
        try {
            amount = Double.parseDouble(amountStr);
            if (amount <= 0) {
                Toast.makeText(this, "El monto debe ser mayor a cero", Toast.LENGTH_SHORT).show();
                return;
            }
        } catch (NumberFormatException e) {
            Toast.makeText(this, "Monto inválido", Toast.LENGTH_SHORT).show();
            return;
        }

        setLoading(true);

        ExpenseEntity expense = new ExpenseEntity(
                description,
                amount,
                CATEGORIES[categoryPosition],
                selectedDate.getTimeInMillis(),
                "user1" // Temporalmente fijo hasta que implementes Auth con Room
        );

        viewModel.addExpense(expense);

        Toast.makeText(this, "¡Gasto creado exitosamente!", Toast.LENGTH_SHORT).show();
        finish();
    }

    private void initSpeechRecognizer() {
        if (SpeechRecognizer.isRecognitionAvailable(this)) {
            speechRecognizer = SpeechRecognizer.createSpeechRecognizer(this);
            speechRecognizer.setRecognitionListener(new RecognitionListener() {
                @Override
                public void onReadyForSpeech(Bundle params) {
                    isListening = true;
                    btnMic.setColorFilter(ContextCompat.getColor(AddExpenseActivity.this, R.color.red_600));
                    Toast.makeText(AddExpenseActivity.this, "Escuchando...", Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onBeginningOfSpeech() {
                }

                @Override
                public void onRmsChanged(float rmsdB) {
                }

                @Override
                public void onBufferReceived(byte[] buffer) {
                }

                @Override
                public void onEndOfSpeech() {
                    isListening = false;
                    btnMic.setColorFilter(ContextCompat.getColor(AddExpenseActivity.this, R.color.gray_400));
                }

                @Override
                public void onError(int error) {
                    isListening = false;
                    btnMic.setColorFilter(ContextCompat.getColor(AddExpenseActivity.this, R.color.gray_400));
                    Toast.makeText(AddExpenseActivity.this, "Error en reconocimiento", Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onResults(Bundle results) {
                    ArrayList<String> matches = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                    if (matches != null && !matches.isEmpty()) {
                        etDescription.setText(matches.get(0));
                    }
                }

                @Override
                public void onPartialResults(Bundle partialResults) {
                }

                @Override
                public void onEvent(int eventType, Bundle params) {
                }
            });
        }
    }

    private void toggleVoiceRecognition() {
        if (isListening) {
            speechRecognizer.stopListening();
        } else {
            if (ContextCompat.checkSelfPermission(this,
                    Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.RECORD_AUDIO },
                        REQUEST_RECORD_AUDIO_PERMISSION);
            } else {
                startListening();
            }
        }
    }

    private void startListening() {
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "es-MX");
        speechRecognizer.startListening(intent);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_RECORD_AUDIO_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startListening();
            } else {
                Toast.makeText(this, "Permiso de audio denegado", Toast.LENGTH_SHORT).show();
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (speechRecognizer != null) {
            speechRecognizer.destroy();
        }
    }

    private void setLoading(boolean loading) {
        isLoading = loading;
        btnSave.setEnabled(!loading);
        btnSave.setText(loading ? "Guardando..." : "Guardar");
        etDescription.setEnabled(!loading);
        etAmount.setEnabled(!loading);
        spinnerCategory.setEnabled(!loading);
        etDate.setEnabled(!loading);
    }
}

File: app/src/main/kotlin/com/gastosapp/AddExpenseActivity.kt
package com.gastosapp

import android.Manifest
import android.app.DatePickerDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.gastosapp.data.local.ExpenseEntity
import com.gastosapp.databinding.ActivityAddExpenseBinding
import com.gastosapp.viewmodel.ExpenseViewModel
import dagger.hilt.android.AndroidEntryPoint
import java.text.SimpleDateFormat
import java.util.*

@AndroidEntryPoint
class AddExpenseActivity : AppCompatActivity() {

    private lateinit var binding: ActivityAddExpenseBinding
    private val viewModel: ExpenseViewModel by viewModels()
    private var speechRecognizer: SpeechRecognizer? = null
    private var isListening = false
    private val selectedDate = Calendar.getInstance()
    private val voiceTextBuilder = StringBuilder()

    companion object {
        private const val REQUEST_RECORD_AUDIO_PERMISSION = 200
        private val CATEGORIES =
                arrayOf(
                        "Selecciona una categoría",
                        "Alimentos",
                        "Transporte",
                        "Entretenimiento",
                        "Salud",
                        "Servicios",
                        "Otros"
                )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddExpenseBinding.inflate(layoutInflater)
        setContentView(binding.root)

        initViews()
        setupCategorySpinner()
        setupDatePicker()
    }

    private fun initViews() {
        binding.btnBack.setOnClickListener { finish() }
        binding.btnSave.setOnClickListener { handleSave() }
        binding.btnMic.setOnClickListener { toggleVoiceRecognition() }

        initSpeechRecognizer()
        updateDateDisplay()
    }

    private fun setupCategorySpinner() {
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, CATEGORIES)
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.spinnerCategory.adapter = adapter
    }

    private fun setupDatePicker() {
        binding.etDate.setOnClickListener {
            DatePickerDialog(
                            this,
                            { _, year, month, dayOfMonth ->
                                selectedDate.set(year, month, dayOfMonth)
                                updateDateDisplay()
                            },
                            selectedDate.get(Calendar.YEAR),
                            selectedDate.get(Calendar.MONTH),
                            selectedDate.get(Calendar.DAY_OF_MONTH)
                    )
                    .show()
        }
        binding.etDate.isFocusable = false
    }

    private fun updateDateDisplay() {
        val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        binding.etDate.setText(sdf.format(selectedDate.time))
    }

    private fun handleSave() {
        val description = binding.etDescription.text.toString().trim()
        val amountStr = binding.etAmount.text.toString().trim()
        val categoryPosition = binding.spinnerCategory.selectedItemPosition
        val userId = intent.getStringExtra("user_id") ?: "default_user"

        if (description.isEmpty()) {
            Toast.makeText(this, "Por favor ingresa una descripción", Toast.LENGTH_SHORT).show()
            return
        }

        if (amountStr.isEmpty()) {
            Toast.makeText(this, "Por favor ingresa un monto", Toast.LENGTH_SHORT).show()
            return
        }

        if (categoryPosition == 0) {
            Toast.makeText(this, "Por favor selecciona una categoría", Toast.LENGTH_SHORT).show()
            return
        }

        val amount =
                amountStr.toDoubleOrNull()
                        ?: run {
                            Toast.makeText(this, "Monto inválido", Toast.LENGTH_SHORT).show()
                            return
                        }

        if (amount <= 0) {
            Toast.makeText(this, "El monto debe ser mayor a cero", Toast.LENGTH_SHORT).show()
            return
        }

        setLoading(true)

        val expense =
                ExpenseEntity(
                        description = description,
                        amount = amount,
                        category = CATEGORIES[categoryPosition],
                        timestamp = selectedDate.timeInMillis,
                        userId = userId
                )

        viewModel.addExpense(expense)
        Toast.makeText(this, "¡Gasto creado exitosamente!", Toast.LENGTH_SHORT).show()
        finish()
    }

    private fun initSpeechRecognizer() {
        if (SpeechRecognizer.isRecognitionAvailable(this)) {
            speechRecognizer =
                    SpeechRecognizer.createSpeechRecognizer(this).apply {
                        setRecognitionListener(
                                object : RecognitionListener {
                                    override fun onReadyForSpeech(params: Bundle?) {
                                        isListening = true
                                        binding.btnMic.setColorFilter(
                                                ContextCompat.getColor(
                                                        this@AddExpenseActivity,
                                                        R.color.red_600
                                                )
                                        )
                                        Toast.makeText(
                                                        this@AddExpenseActivity,
                                                        "Escuchando...",
                                                        Toast.LENGTH_SHORT
                                                )
                                                .show()
                                    }

                                    override fun onBeginningOfSpeech() {}
                                    override fun onRmsChanged(rmsdB: Float) {}
                                    override fun onBufferReceived(buffer: ByteArray?) {}
                                    override fun onEndOfSpeech() {
                                        if (isListening) startListening()
                                    }

                                    override fun onError(error: Int) {
                                        if (isListening) startListening()
                                        else {
                                            isListening = false
                                            binding.btnMic.setColorFilter(
                                                    ContextCompat.getColor(
                                                            this@AddExpenseActivity,
                                                            R.color.gray_400
                                                    )
                                            )
                                        }
                                    }

                                    override fun onResults(results: Bundle?) {
                                        val matches =
                                                results?.getStringArrayList(
                                                        SpeechRecognizer.RESULTS_RECOGNITION
                                                )
                                        if (!matches.isNullOrEmpty()) {
                                            val result = matches[0]
                                            if (voiceTextBuilder.isNotEmpty())
                                                    voiceTextBuilder.append(" ")
                                            voiceTextBuilder.append(result)
                                            binding.etDescription.setText(
                                                    voiceTextBuilder.toString()
                                            )
                                        }
                                        if (isListening) startListening()
                                    }

                                    override fun onPartialResults(partialResults: Bundle?) {}
                                    override fun onEvent(eventType: Int, params: Bundle?) {}
                                }
                        )
                    }
        }
    }

    private fun toggleVoiceRecognition() {
        if (isListening) {
            isListening = false
            speechRecognizer?.stopListening()
            binding.btnMic.setColorFilter(ContextCompat.getColor(this, R.color.gray_400))
        } else {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) !=
                            PackageManager.PERMISSION_GRANTED
            ) {
                ActivityCompat.requestPermissions(
                        this,
                        arrayOf(Manifest.permission.RECORD_AUDIO),
                        REQUEST_RECORD_AUDIO_PERMISSION
                )
            } else {
                voiceTextBuilder.setLength(0)
                isListening = true
                startListening()
            }
        }
    }

    private fun startListening() {
        val intent =
                Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
                    putExtra(
                            RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                            RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
                    )
                    putExtra(RecognizerIntent.EXTRA_LANGUAGE, "es-MX")
                }
        speechRecognizer?.startListening(intent)
    }

    override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<out String>,
            grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_RECORD_AUDIO_PERMISSION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startListening()
            } else {
                Toast.makeText(this, "Permiso de audio denegado", Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        speechRecognizer?.destroy()
    }

    private fun setLoading(loading: Boolean) {
        binding.btnSave.isEnabled = !loading
        binding.btnSave.text = if (loading) "Guardando..." else "Guardar"
        binding.etDescription.isEnabled = !loading
        binding.etAmount.isEnabled = !loading
        binding.spinnerCategory.isEnabled = !loading
        binding.etDate.isEnabled = !loading
    }
}
